Bluetooth: tester: Add tester application with MMDL support

michal-narajowski opened the pull request:
Add support for automatic testing of Mesh Models. This app is ported
from Zephyr repository. It should only be used with MMDL test cases.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>
Signed-off-by: Agata Ponitka <agata.ponitka@codecoup.pl>

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-3883
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

trond-snekvik commented on tests/bluetooth/tester/src/gap.c:
```diff
@@ -0,0 +1,269 @@
+/* gap.c - Bluetooth GAP Tester */
```
Are the GAP commands necessary to run the MMDL test suite?

michal-narajowski replied:
I left only what was necessary without moving the code around. This way there is no need to modify autopts code. In `gap_init()` we call `bt_enable()` and in `gap_read_ctrl_info()` we set an identity address and communicate it to PTS. This is how it was designed and while I think it would be nice to change it, we would have to do this in upstream (decouple gap and mesh).

trond-snekvik replied:
Alright, makes sense. Keep it :)

trond-snekvik commented on tests/bluetooth/tester/src/bttester.c:
```diff
@@ -0,0 +1,292 @@
+/* bttester.c - Bluetooth Tester */
```
I think this file can be pulled out into an upstream subsys/bluetooth/qualification module, along with common test modules, like GAP, GATT, Mesh and MMDL perhaps. This way, it can be included in test applications in end product verification, as we talked about on Monday. Let's get this PR merged first though.

cc @joerchan @carlescufi

carlescufi replied:
Yes, I completely agree @trond-snekvik. We should get this upstream, since we already have the equivalent for regular host, and this only extends it with additional features.

trond-snekvik commented on tests/bluetooth/tester/src/mmdl.c:
```diff
@@ -0,0 +1,4320 @@
+#include <bluetooth/bluetooth.h>
```
Missing license header.

trond-snekvik commented on tests/bluetooth/tester/src/model_handler.c:
```diff
@@ -0,0 +1,608 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
```
Should be `SPDX-License-Identifier: LicenseRef-Nordic-5-Clause`, in accordance with #3842

trond-snekvik commented on tests/bluetooth/tester/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+  bluetooth.general.tester:
+    platform_allow: qemu_x86
+    tags: bluetooth
+    harness: bluetooth
```
Could you please add an additional target for building on nrf52840, just so the tests can be compiled as part of CI?

ie
```yml
  bluetooth.general.tester.build:
    build_only: true
    platform_allow: nrf52840dk_nrf52840
    tags: bluetooth ci_build
```

trond-snekvik reviewed (changes requested):
This is great! I think there are a couple of obvious candidates for "test module" files in the tester app, and this PR really highlights the need. I don't think we should do that work as part of this though, this code has been in a PR state for way too long.

A couple of nitpicks, but the test code looks great.

trond-snekvik commented:
Oh, and I think this should get its own code owners entry, I suggest
```
/tests/bluetooth/tester                        @joerchan @carlescufi @trond-snekvik
```

The test name should be /tests/bluetooth/qualification, or /tests/bluetooth/pts.

michal-narajowski commented:
@trond-snekvik I think I addressed all of your comments. Not sure about the last one: 
```
The test name should be /tests/bluetooth/qualification, or /tests/bluetooth/pts.
```

trond-snekvik commented:
@carlescufi @joerchan Any preferences on the test name? See the previous comment from @michal-narajowski 

Otherwise, this looks good to me.

michal-narajowski commented:
I removed the Prov Enable BTP command as it won't be necessary once https://github.com/zephyrproject-rtos/zephyr/pull/32369 is merged.
